mypy/checker.py

```diff
@@ -1839,18 +1839,39 @@ def exception_type(self, n: Node) -> Type:
         self.fail('Unsupported exception', n)
         return AnyType()
 
+    def check_for_function_like_exception_type(self, type: Type, context: Context) -> Type:
```

Comment:
Because this can return None and not only an actual `Type`, let's annotate as `Optional[Type]`. PEP 484 specifies that types don't implicitly contain None; we don't quite yet implement that properly in mypy, so we don't catch this, but there's an active PR for it (#1562) and we will soon.